### PR TITLE
Interpreter: allow inspecting block vars without affecting program

### LINKED ISSUE #12518: [Interpreter mode] Integer overflow when do each loop.
Code here:

```crystal
class InsertionSort
  def self.sort(arr)
    (1..arr.size - 1).each do |i|
      (0..i - 1).each do |j|
        debugger
        arr[j], arr[i] = arr[i], arr[j] if arr[i] < arr[j]
      end
    end
  end
end

arr = [4, 8, 7, 2, 6, 5, 3, 0, 9, 1]

InsertionSort.sort(arr)
p arr
```

```
 ╰─ $ 1  cr i 1.cr 
From: 1.cr:21:44 InsertionSort.class#sort:

    16: class InsertionSort
    17:   def self.sort(arr)
    18:     (1..arr.size - 1).each do |i|
    19:       (0..i - 1).each do |j|
    20:         debugger
 => 21:         arr[j], arr[i] = arr[i], arr[j] if arr[i] < arr[j]
    22:       end
    23:     end
    24:   end
    25: end
    26: 

pry(InsertionSort.sort)> i
=> 632700968
pry(InsertionSort.sort)> j
=> 0
```

After typing Ctrl + d

```
pry(InsertionSort.sort)> Invalid memory access (signal 11) at address 0x8
[0x556a75ecfa06] *Exception::CallStack::print_backtrace:Nil +118 in crystal
[0x556a75ea5776] ~procProc(Int32, Pointer(LibC::SiginfoT), Pointer(Void), Nil) +310 in crystal
[0x7f782ee51a00] ?? +140154159569408 in /usr/lib/libc.so.6
[0x7f782ef6c804] ?? +140154160728068 in /usr/lib/libc.so.6
[0x556a75ec3997] *Pointer(UInt8) +103 in crystal
[0x556a75ec3921] *Pointer(UInt8) +17 in crystal
[0x556a75ec3901] *Pointer(UInt8) +17 in crystal
[0x556a76ddd27f] *Crystal::Repl::Interpreter#interpret<Crystal::ASTNode+, Crystal::Type+>:Crystal::Repl::Value +388687 in crystal
[0x556a76d7da1c] *Crystal::Repl::Interpreter#interpret<Crystal::ASTNode+, Hash(String, Crystal::MetaVar)>:Crystal::Repl::Value +2284 in crystal
[0x556a7735102d] *Crystal::Repl#interpret<Crystal::Expressions>:Crystal::Repl::Value +109 in crystal
[0x556a77350d66] *Crystal::Repl#interpret_and_exit_on_error<Crystal::Expressions>:Crystal::Repl::Value +38 in crystal
[0x556a77351832] *Crystal::Repl#run_file<String, Array(String)>:Crystal::Repl::Value +194 in crystal
[0x556a772910cb] *Crystal::Command#repl:(Crystal::Repl::Value | Nil) +1451 in crystal
[0x556a7728b9b1] *Crystal::Command#run:(Bool | Crystal::Repl::Value | Nil) +1393 in crystal
[0x556a7728b2bb] *Crystal::Command::run<Array(String)>:(Bool | Crystal::Repl::Value | Nil) +43 in crystal
[0x556a7728b22e] *Crystal::Command::run:(Bool | Crystal::Repl::Value | Nil) +46 in crystal
[0x556a75e80909] __crystal_main +2457 in crystal
[0x556a760747c3] *Crystal::main_user_code<Int32, Pointer(Pointer(UInt8))>:Nil +19 in crystal
[0x556a7607472a] *Crystal::main<Int32, Pointer(Pointer(UInt8))>:Int32 +58 in crystal
[0x556a75e8dec6] main +6 in crystal
[0x7f782ee3c290] ?? +140154159481488 in /usr/lib/libc.so.6
[0x7f782ee3c34a] __libc_start_main +138 in /usr/lib/libc.so.6
[0x556a75e7fe95] _start +37 in crystal
[0x0] ???
```

```
 ╰─ $ 11  cr version
Crystal 1.6.0-dev [c502c77d8] (2022-09-16)

LLVM: 14.0.6
Default target: x86_64-pc-linux-gnu
```

### PR DESCRIPTION
**Pull request by asterite**
Fixes #12518

This would would also be good for 1.6

### CODE REVIEW

**Comment by asterite** [comment]
I also fixed another bug that I noticed that happened when you introduced new vars while debugging inside a block.

So now the only thing remaining is being able to define new vars inside a block and later change their type (similar to what you can do at the top level,) but I'll leave that for another PR.

**Comment by asterite** [comment]
Here's an example session showing how it works now:

```crystal
✗ bin/crystal i foo.cr
Using compiled compiler at .build/crystal
From: foo.cr:6:44 InsertionSort.class#sort:

     1: class InsertionSort
     2:   def self.sort(arr)
     3:     (1..arr.size - 1).each do |i|
     4:       (0..i - 1).each do |j|
     5:         debugger
 =>  6:         arr[j], arr[i] = arr[i], arr[j] if arr[i] < arr[j]
     7:       end
     8:     end
     9:   end
    10: end
    11:

pry(InsertionSort.sort)> i
=> 1
pry(InsertionSort.sort)> j
=> 0
pry(InsertionSort.sort)> arr
=> [4, 8, 7, 2, 6, 5, 3, 0, 9, 1]
pry(InsertionSort.sort)> b = [1, 2, 3]
=> [1, 2, 3]
pry(InsertionSort.sort)> b
=> [1, 2, 3]
pry(InsertionSort.sort)> b.reverse
=> [3, 2, 1]
pry(InsertionSort.sort)> continue
From: foo.cr:6:44 InsertionSort.class#sort:

     1: class InsertionSort
     2:   def self.sort(arr)
     3:     (1..arr.size - 1).each do |i|
     4:       (0..i - 1).each do |j|
     5:         debugger
 =>  6:         arr[j], arr[i] = arr[i], arr[j] if arr[i] < arr[j]
     7:       end
     8:     end
     9:   end
    10: end
    11:

pry(InsertionSort.sort)> i
=> 2
pry(InsertionSort.sort)> j
=> 0
pry(InsertionSort.sort)> continue
From: foo.cr:6:44 InsertionSort.class#sort:

     1: class InsertionSort
     2:   def self.sort(arr)
     3:     (1..arr.size - 1).each do |i|
     4:       (0..i - 1).each do |j|
     5:         debugger
 =>  6:         arr[j], arr[i] = arr[i], arr[j] if arr[i] < arr[j]
     7:       end
     8:     end
     9:   end
    10: end
    11:

pry(InsertionSort.sort)> ^C
```

**Comment by zw963** [comment]
Add to milestone 1.6 ?

File: src/compiler/crystal/interpreter/interpreter.cr [hunk 1]
```diff
@@ -167,8 +167,10 @@ class Crystal::Repl::Interpreter
     compiled_def = @compiled_def
 
     # Declare or migrate local variables
-    if !compiled_def || in_pry
-      migrate_local_vars(@local_vars, meta_vars)
+    # TODO: we should also migrate variables if we are outside of a block
+    # in a pry session, but that's tricky so we'll leave it for later.
+    if (!compiled_def || in_pry)
```
Comment (straight-shoota) [inline src/compiler/crystal/interpreter/interpreter.cr:172]: Is there a particular reason for adding these parentheses?

**Review by straight-shoota — APPROVED** [review]

File: src/compiler/crystal/interpreter/interpreter.cr (hunk 1 above)
Comment (asterite) [reply src/compiler/crystal/interpreter/interpreter.cr:172]: Ah, ko, it's from a previous change that I later undid, but apparently not fully.

Comment (straight-shoota) [reply src/compiler/crystal/interpreter/interpreter.cr:172]:
```suggestion
    if !compiled_def || in_pry
```

**Pull request merged**